co-log/src/Colog/Actions.hs

```diff
@@ -92,9 +102,64 @@ withLogRotation
     -> (LogAction m msg -> IO r)
     -- ^ Continuation action
     -> IO r
-withLogRotation sizeLimit filesLimit path cleanup mkAction cont = cont rotationAction
+withLogRotation sizeLimit filesLimit path cleanup mkAction cont = do
+  handleRef <- bracket (openFile path AppendMode) hClose newIORef
+  cont $ rotationAction handleRef
   where
-    rotationAction :: LogAction m msg
-    rotationAction = LogAction $ \msg -> do
-        withFile path AppendMode writeFileLoop
--}
+    rotationAction refHandle
+      = LogAction $ \msg -> do
+          oldFiles <- getOldFiles filesLimit path
+          mapM_ (liftIO . cleanup) oldFiles
+          handle <- readIORef refHandle
+          sizeLimitReached <- isFileSizeLimitReached sizeLimit handle
+
+          if sizeLimitReached then do
```

Comment:
Actually, the logic can be slightly simpler. It's okay to always write `msg` to the existing `Handle` and perform renaming only after message is written. Like this:

```haskell
outputMsg msg handle
whenM (isFileSizeLimitReached sizeLimit handle) cleanupAndRotate
```

Also, you need to cleanup only when you perform logger rotation. Don't need to cleanup every time on each message.